server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java

```diff
@@ -1439,6 +1439,7 @@ private void addSink(SegmentIdWithShardSpec identifier, Sink sink)
         sink.getVersion(),
         identifier.getShardSpec().createChunk(sink)
     );
+    ((SinkQuerySegmentWalker) texasRanger).registerUpgradedPendingSegment(identifier, identifier);
```

Comment:
Here, we effectively add an item to the `upgradeDescriptorTimeline`. So we should also remove this item when this class calls `sinkTimeline.remove`.

Reply:
I believe that every registered version of the sink must be removed from the upgradedSegmentTimeline in such a case

Reply:
Yes, we would need to remove the base as well as the upgraded segments from the timeline.